File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 Approach: Use two pointer to find a subarray with no repeating elements, as element gets repeated move left pointer until duplicate element is no longer included in subarray. Use set to remember seen elements.
 TC= O(n)
 SC= O(n)
 */
var lengthOfLongestSubstring = function(s) {
    const set= new Set();
    var result= 0, j= 0;
    for(let i=0; i<s.length; i++){
        while(set.has(s[i])){
            set.delete(s[j]);
            j++;
        }
        set.add(s[i]);
        result= Math.max(result, set.size);
    }
    return result;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        mem = defaultdict(int)
        ans = 0
        for r in range(len(s)):
            ch = s[r]
            mem[ch] += 1
            while mem[ch] > 1:
                mem[s[left]] -= 1
                left += 1
            ans = max(ans, r - left + 1)
        
        return ans

# TC: O(n)
# SC: O(1)
# Clarification: 
# Brute Force:
# Approach: 
# Testing: